StereoKit/Systems/Renderer.cs

```diff
@@ -284,6 +291,73 @@ public static void Screenshot(Vec3 from, Vec3 at, int width, int height, string
+		/// </param>
+		/// <param name="fieldOfViewDegrees">The angle of the viewport, in 
+		/// degrees.</param>
+		public static void Screenshot(ScreenshotCallback onScreenshot, Vec3 from, Vec3 at, int width, int height, float fieldOfViewDegrees = 90)
+		{
+			if (_renderCaptureCallbacks is null) _renderCaptureCallbacks = new ConcurrentQueue<RenderOnScreenshotCallback>();
+			RenderOnScreenshotCallback renderCaptureCallback = (IntPtr dataPtr, int w, int h, IntPtr context) =>
+			{
+				if (_renderCaptureCallbacks.TryDequeue(out _))
+					onScreenshot.Invoke(dataPtr, w, h);
+			};
+			_renderCaptureCallbacks.Enqueue(renderCaptureCallback);
+			NativeAPI.render_screenshot_capture(renderCaptureCallback, from, at, width, height, fieldOfViewDegrees);
+		}
+
+		/// <summary>Schedules a screenshot for the end of the frame! The view
+		/// will be rendered from the given position at the given point, with a
+		/// resolution the same size as the screen's surface. This overload
+		/// allows for retrieval of the color data directly from the render
+		/// thread! You can use the color data directly by saving/processing it
+		/// inside your callback, or you can keep the data alive for as long as
+		/// it is referenced.</summary>
+		/// <param name="onScreenshot">Outputs a reference to the color data
+		/// and its length which represent the current scene from a requested
+		/// viewpoint.</param>
+		/// <param name="camera">A TRS matrix representing the location and
+		/// orientation of the camera. This matrix gets inverted later on, so
+		/// no need to do it yourself.</param>
+		/// <param name="projection">The projection matrix describes how the
+		/// geometry is flattened onto the draw surface. Normally, you'd use 
+		/// Matrix.Perspective, and occasionally Matrix.Orthographic might be
+		/// helpful as well.</param>
+		/// <param name="layerFilter">This is a bit flag that allows you to
+		/// change which layers StereoKit renders for this particular render
+		/// viewpoint. To change what layers a visual is on, use a Draw
+		/// method that includes a RenderLayer as a parameter.</param>
+		/// <param name="clear">Describes if and how the rendertarget should
+		/// be cleared before rendering. Note that clearing the target is
+		/// unaffected by the viewport, so this will clean the entire 
+		/// surface!</param>
+		public static void Screenshot(ScreenshotCallback onScreenshot, Matrix camera, Matrix projection, int width, int height, RenderLayer layerFilter = RenderLayer.All, RenderClear clear = RenderClear.All, Rect viewport = default(Rect))
+		{
+			if (_renderCaptureCallbacks is null) _renderCaptureCallbacks = new ConcurrentQueue<RenderOnScreenshotCallback>();
+			RenderOnScreenshotCallback renderCaptureCallback = (IntPtr dataPtr, int w, int h, IntPtr context) =>
+			{
+				if (_renderCaptureCallbacks.TryDequeue(out _))
+					onScreenshot.Invoke(dataPtr, w, h);
+			};
+			_renderCaptureCallbacks.Enqueue(renderCaptureCallback);
```

Comment:
This enqueue dequeue thing feels a little shaky to me! The only thing that holds this to be correct is that both the `ConcurrentQueue` and the underlying processing of screenshots are independently FIFO. I think it's still the right data structure, but it might be good to note that this is the case in the comments of `render_check_screenshots`, so if we change the behavior of callbacks in the future to be more async, we know to still preserve the FIFO behavior of the code.

Reply:
Having an async callback sounds interesting. It sorta just fires them all natively and the enqueued user-defined callbacks can be done in any order. Hmm maybe a list with a lock is the way to go after all, in order to finalize the completed actions

But without async callbacks, it's pretty much FIFO. I guess it could change from ConcurrentQueue to Queue since it's not actually thread safe, as the current C# enqueue and native C list add are mutually exclusive without a lock

Reply:
Oh! What if we pass in another callback natively that essentially says "# of screenshots are complete" and this can be used by C# to dequeue that many times, or any C API wanting to know how many of the screenshots were just processed?

Reply:
I think that even if this becomes more async, it will still likely be FIFO? I believe the GPU should still do the work in order, so it shouldn't be much of a burden to just ensure that actually happens and promise it to the developer. Having that promise seems valuable anyhow, like if I wanted to screenshot a series of images for a gif, having them arrive out-of-order would just complicate matters.

But yeah, I guess the native list isn't threadsafe, so the C# side being safe there doesn't make it much better! But `ConcurrentQueue` _shouldn't_ add much overhead, so feel free to leave it in! From what I've read, it's implemented with atomic operations rather than locks. I've not so long ago figured how to do some atomic stuff in C, so at some point later on I'd like to revise SK's builtin data structures to utilize that.

Reply:
I've thought about having a poll based alternative for anything that uses callbacks? I think that'd probably be a lot more "StereoKit idiomatic", but determining the source of things in a poll queue might be a little problematic here! I don't think I'm quite going to do this though.

Reply:
I changed it back to a `Queue` mainly so I remember to call the screenshot from SK's main thread (via a stepper or `SK.ExecuteOnMain`). Then tested it from a separate thread and it worked great without the added concurrency. [I don't think a ConcurrentQueue is too relevant for this scenario](https://learn.microsoft.com/en-us/dotnet/standard/collections/thread-safe/when-to-use-a-thread-safe-collection#concurrentqueuet-vs-queuet), with the FIFO order guaranteed. I'll also add a FIFO comment in there then ready for review again.

I guess it'd be like polling for OpenXR events but SK style.

Reply:
> Having that promise seems valuable anyhow, like if I wanted to screenshot a series of images for a gif, having them arrive out-of-order would just complicate matters.

Huh now I want a script to make speedy gifs for sharing! :D

Reply:
That was kinda the idea with the RenderCamera script! I just use ScreenToGif for converting the bmp frames into the final gif :) I'm sure there's a NuGet out there that could handle gif stuff instead, but I didn't want to worry about that for a single file tool. I had been thinking about adding more direct gif support at one point ([this library](https://github.com/notnullnotvoid/msf_gif) in particular looked super nice), and now that I'm peeking back at it I'm not too sure why I didn't!

Reply:
I'm a big fan of changing the perspectives in RenderCamera. Woa msf_gif looks so clean and writes a frame at a time to spare the memory! A few of their [comments](https://github.com/notnullnotvoid/msf_gif/blob/7829c8f439d013deeb54eba94761403e1da2a960/msf_gif.h#L318) made me chuckle. As a single header with RGBA format by default, this could be a real winner